### DISCUSSION TITLE
Dynamically define cfengine class functions

### DISCUSSION THREAD

**Discussion by basvandervlies** [cfengine/core#4891] [Q&A]
I have asked this question in the newsgroup, maybe I missed the answer!

I want to use the cfengine class functions dynamically. I will try to explain it, eg:
```
"class_set" expression => isdir("/etc"); 
```

But in the example above the `isdir("/etc")` is a variable defined in a json file
```
"class_set" expression => $(def.class_func)
```

also tried:
```
reports:
  "It works"
      if => or( $def.class_func );
```
 
Is this possible?

**Comment by olehermanse** [comment]
@basvandervlies 

The only way I can think of is to render a policy file, from a template, with the correct function call(s).

We could consider adding another function:

```cf3
"class_set" expression => call_function("isdir", "/etc");
```

What do you think about that idea?

**Reply by basvandervlies** [reply]
@olehermanse and when a function has multuiple args?  eg: `classmatch(regex, tag1, tag2, ...)` 
I was more just pass the fucntion string to the `call_function`, eg:
 * `"class_set" expression => call_function('isdir("/etc")');`

Or is the above syntax problematic?

**Reply by olehermanse** [reply]
@basvandervlies It could be done, but it does introduce more complexity, as you have to parse CFEngine syntax inside that string. Parsing is always difficult and adds complexity, i.e.:

```
"class_set" expression => call_function('isdir()');
"class_set" expression => call_function('isdir("")');
"class_set" expression => call_function('isdir(blah)');
"class_set" expression => call_function('isdir("blah")');
"class_set" expression => call_function('isdir(`blah`)');
"class_set" expression => call_function('isdir(${myvar})');
"class_set" expression => call_function('isdir("${myvar}")');
"class_set" expression => call_function('is${funcname}()');
"class_set" expression => call_function('string_replace(isdir("/etc"), "any", "yes"');
```


It also adds a lot of quoting inside of quoted strings, have to be careful with alternating types of quots or escape sequences.

etc.

**Reply by olehermanse** [reply]
@basvanderlies I would make call_function take n arguments, where the last n-1 of them are sent to the called function.
